framework/src/variables/MooseVariableFV.C

```diff
@@ -734,28 +723,30 @@ MooseVariableFV<OutputType>::adGradSln(const FaceInfo & fi, const bool correct_s
   mooseError("MooseVariableFV::adGradSln only supported for global AD indexing");
 #endif
 
-  // Use a pointer to choose the right reference
-  auto face_grad = uncorrectedAdGradSln(fi, correct_skewness);
+  bool internal_face = isInternalFace(fi);
+
+  bool var_defined_on_elem = fi.varDefinedOnElem(this->name());
+  const Elem * const elem = &fi.elem();
+  const Elem * const neighbor = fi.neighborPtr();
 
-  auto tup = Moose::FV::determineElemOneAndTwo(fi, *this);
-  const Elem * const elem_one = std::get<0>(tup);
-  const Elem * const elem_two = std::get<1>(tup);
-  const bool elem_is_elem_one = std::get<2>(tup);
-  mooseAssert(elem_is_elem_one ? elem_one == &fi.elem() && elem_two == fi.neighborPtr()
-                               : elem_one == fi.neighborPtr() && elem_two == &fi.elem(),
-              "The determineElemOneAndTwo utility got the elem_is_elem_one value wrong.");
+  const ADReal side_one_value =
+      var_defined_on_elem ? getElemValue(elem) : getNeighborValue(elem, fi);
+  const ADReal side_two_value =
+      var_defined_on_elem ? getNeighborValue(neighbor, fi) : getElemValue(neighbor);
 
-  const ADReal elem_one_value = getElemValue(elem_one);
-  const ADReal elem_two_value = getNeighborValue(elem_two, fi, elem_one_value);
-  const ADReal & elem_value = elem_is_elem_one ? elem_one_value : elem_two_value;
-  const ADReal & neighbor_value = elem_is_elem_one ? elem_two_value : elem_one_value;
+  const Real delta = internal_face ? std::abs(fi.dCN() * fi.normal())
+                                   : fi.cellCenterToFaceDistance(var_defined_on_elem);
 
-  // perform the correction. Note that direction is important here because we have a minus sign.
-  // Neighbor has to be neighbor, and elem has to be elem. Hence all the elem_is_elem_one logic
-  // above
-  face_grad += ((neighbor_value - elem_value) / fi.dCNMag() - face_grad * fi.eCN()) * fi.eCN();
+  auto face_grad = ((side_two_value - side_one_value) / delta);
+
+  // We only need nonorthogonal correctors in 2+ dimensions
+  if (_mesh.dimension() > 1)
+  {
+    const auto & interpolated_gradient = uncorrectedAdGradSln(fi, correct_skewness);
+    face_grad += interpolated_gradient * (fi.normal() - fi.dCN() / delta);
+  }
 
-  return face_grad;
+  return face_grad * fi.normal();
```

Comment:
This might need refinement to get the correct gradient. Now it focuses to get the surface normal gradient.